semgrep-core/src/parsing/other/yaml_to_generic.ml

```diff
@@ -313,8 +313,17 @@ let parse (env : env) : G.expr list =
         let rest, end_pos = read_sequence acc in
         (seq :: rest, end_pos)
   and read_mappings acc : G.expr list * E.pos =
+    (* Capture the mutable `last_event` field prior to parsing, as the *)
+    (* `Mapping_end` event needs to make use of it. *)
+    let last_ev = env.last_event in
```

Comment:
I think I would rather grab the last element of the accumulator (if it exists), and use the position of that element. But I get a strange typecheck error when I do so, and my ocaml-fu is inadequate for solving it 
```ocaml
    | E.Mapping_end, pos ->
        (* The `Mapping_end` event position is inaccurate: it specifies the *)
        (* position immediately prior to the next token. This causes comments *)
        (* and the leading whitespace of the next yaml element to be chomped on *)
        (* autofix. Instead, take the position of the previous element, which *)
        (* is either the last element in the mapping, or the start of the map *)
        (* if empty. *)
        (match List.rev acc with
         | [] -> (acc, pos)
         | (_, last_pos) :: _ -> (acc, last_pos))
```

```
Line 325, characters 18-21:
325 |          | [] -> (acc, pos)
                        ^^^
Error: This expression has type ('a * 'b) list
       but an expression was expected of type G.expr list
       Type 'a * 'b is not compatible with type G.expr = AST_generic.expr 

```